context-propagation/src/main/java/io/micrometer/context/DefaultContextSnapshot.java

```diff
@@ -83,12 +83,17 @@ public Scope setThreadLocals(Predicate<Object> keyPredicate) {
     }
 
     @SuppressWarnings("unchecked")
-    private static <V> Map<Object, Object> setThreadLocal(Object key, V value, ThreadLocalAccessor<?> accessor,
-            @Nullable Map<Object, Object> previousValues) {
+    private static <V> Map<Object, Object> setThreadLocal(Object key, @Nullable V value,
+            ThreadLocalAccessor<?> accessor, @Nullable Map<Object, Object> previousValues) {
 
         previousValues = (previousValues != null ? previousValues : new HashMap<>());
         previousValues.put(key, accessor.getValue());
-        ((ThreadLocalAccessor<V>) accessor).setValue(value);
+        if (value != null) {
+            ((ThreadLocalAccessor<V>) accessor).setValue(value);
+        }
+        else {
+            accessor.resetToSetValue();
```

Comment:
This is a breaking change. Fortunately, it currently won't be called with null as the value, but I'd delay adding this code for safety and have it added when we have the surrounding configurable API to control this behaviour.

Reply:
Actually, it can be called from `setThreadLocals(Predicate<Object> keyPredicate)` and it can provide `null` to `setValue`

Reply:
The existing API doesn't hold some of the promises. The non-null-argument methods can in fact receive null arguments, as demonstrated by the following branch: https://github.com/micrometer-metrics/context-propagation/compare/non-null-api?expand=1

Reply:
#102 improves this scenario.